src/simulate-swap.ts

```diff
@@ -0,0 +1,927 @@
+  }
+  if (tickIndexAbs & 0x4n) {
+    sqrtPrice = (sqrtPrice * 1000200010000n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x8n) {
+    sqrtPrice = (sqrtPrice * 1000400060004n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x10n) {
+    sqrtPrice = (sqrtPrice * 1000800280056n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x20n) {
+    sqrtPrice = (sqrtPrice * 1001601200560n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x40n) {
+    sqrtPrice = (sqrtPrice * 1003204964963n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x80n) {
+    sqrtPrice = (sqrtPrice * 1006420201726n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x100n) {
+    sqrtPrice = (sqrtPrice * 1012881622442n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x200n) {
+    sqrtPrice = (sqrtPrice * 1025929181080n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x400n) {
+    sqrtPrice = (sqrtPrice * 1052530684591n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x800n) {
+    sqrtPrice = (sqrtPrice * 1107820842005n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x1000n) {
+    sqrtPrice = (sqrtPrice * 1227267017980n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x2000n) {
+    sqrtPrice = (sqrtPrice * 1506184333421n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x4000n) {
+    sqrtPrice = (sqrtPrice * 2268591246242n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x8000n) {
+    sqrtPrice = (sqrtPrice * 5146506242525n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x00010000n) {
+    sqrtPrice = (sqrtPrice * 26486526504348n) / FIXED_POINT_DENOMINATOR
+  }
+  if (tickIndexAbs & 0x00020000n) {
+    sqrtPrice = (sqrtPrice * 701536086265529n) / FIXED_POINT_DENOMINATOR
+  }
+
+  if (tickIndex >= 0n) {
+    return rescale(sqrtPrice, FIXED_POINT_SCALE, SQRT_PRICE_SCALE)
+  } else {
+    let sqrtPriceInFixedPointScale = (FIXED_POINT_DENOMINATOR * FIXED_POINT_DENOMINATOR) / sqrtPrice
+    return rescale(sqrtPriceInFixedPointScale, FIXED_POINT_SCALE, SQRT_PRICE_SCALE)
+  }
+}
+
+const rescale = (fromValue: bigint, fromScale: bigint, expectedScale: bigint): bigint => {
+  if (expectedScale > fromScale) {
+    const multiplierScale = expectedScale - fromScale
+    return fromValue * 10n ** multiplierScale
+  } else {
+    const denominatorScale = fromScale - expectedScale
+    return fromValue / 10n ** denominatorScale
+  }
+}
+
+const divToTokenUp = (nominator: bigint, denominator: bigint): bigint => {
+  let result = nominator * SQRT_PRICE_DENOMINATOR
+  result += denominator - 1n
+  result /= denominator
+  result += SQRT_PRICE_DENOMINATOR - 1n
+  result /= SQRT_PRICE_DENOMINATOR
+  return result
+}
+
+const divToToken = (nominator: bigint, denominator: bigint): bigint => {
+  let result = nominator * SQRT_PRICE_DENOMINATOR
+  result /= denominator
+  result /= SQRT_PRICE_DENOMINATOR
+  return result
+}
+
+const mulDiv = (a: bigint, b: bigint, denominator: bigint): bigint => {
```

Comment:
It is important that denominator refers strictly to `b` value, so naming it `bDenominator` looks much better